flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CollectionFunctionsITCase.java

```diff
@@ -479,4 +480,165 @@ private Stream<TestSetSpec> arrayUnionTestCases() {
+                                DataTypes.ARRAY(DataTypes.ARRAY(DataTypes.INT())),
+                                DataTypes.ARRAY(DataTypes.ROW(DataTypes.DATE())),
+                                DataTypes.ARRAY(DataTypes.BOOLEAN()),
+                                DataTypes.ARRAY(DataTypes.ROW(DataTypes.BOOLEAN())),
+                                DataTypes.ARRAY(DataTypes.ROW(DataTypes.INT())),
+                                DataTypes.INT().notNull(),
+                                DataTypes.DOUBLE().notNull(),
+                                DataTypes.STRING().notNull(),
+                                DataTypes.ARRAY(DataTypes.ARRAY(DataTypes.INT())),
+                                DataTypes.ARRAY(DataTypes.DATE()))
+                        .testResult($("f0").arrayMax(), "ARRAY_MAX(f0)", 2, DataTypes.INT())
+                        .testResult($("f1").arrayMax(), "ARRAY_MAX(f1)", null, DataTypes.INT())
+                        .testResult($("f2").arrayMax(), "ARRAY_MAX(f2)", 8.0, DataTypes.DOUBLE())
+                        .testResult($("f3").arrayMax(), "ARRAY_MAX(f3)", "def", DataTypes.STRING())
+                        .testResult(
+                                $("f15").arrayMax(),
+                                "ARRAY_MAX(f15)",
+                                LocalDate.of(2092, 7, 19),
+                                DataTypes.DATE())
+                        .testSqlValidationError(
+                                "ARRAY_MAX(f4)",
+                                "SQL validation failed. Invalid function call:\n"
+                                        + "ARRAY_MAX(ARRAY<ROW<`f0` BOOLEAN, `f1` DATE>>")
+                        .testTableApiValidationError(
+                                $("f4").arrayMax(),
+                                "Invalid function call:\n"
+                                        + "ARRAY_MAX(ARRAY<ROW<`f0` BOOLEAN, `f1` DATE>>")
+                        .testSqlValidationError(
+                                "ARRAY_MAX(f5)",
+                                "SQL validation failed. Invalid function call:\n"
+                                        + "ARRAY_MAX(ARRAY<MAP<INT, STRING>>")
+                        .testTableApiValidationError(
+                                $("f5").arrayMax(),
+                                "Invalid function call:\n" + "ARRAY_MAX(ARRAY<MAP<INT, STRING>>)")
+                        .testSqlValidationError(
+                                "ARRAY_MAX(f6)",
+                                "SQL validation failed. Invalid function call:\n"
+                                        + "ARRAY_MAX(ARRAY<ARRAY<INT>>)")
+                        .testTableApiValidationError(
+                                $("f6").arrayMax(),
+                                "Invalid function call:\n" + "ARRAY_MAX(ARRAY<ARRAY<INT>>)")
+                        .testSqlValidationError(
```

Comment:
Yes, it seem that we cannot compare constructed types. Because  it is unable to create an expression evaluator for expression
```

Caused by: org.apache.flink.table.planner.codegen.CodeGenException: Comparable type expected, but was 'ROW<`f0` BOOLEAN, `f1` DATE> NOT NULL'.
	at org.apache.flink.table.planner.codegen.CodeGenUtils$.requireComparable(CodeGenUtils.scala:399)
```

and because it will call this function   
```
public static boolean isComparable(LogicalType type) {
        return !isRaw(type) && !isMap(type) && !isMultiset(type) && !isRow(type) && !isArray(type);
    }
```

so we cannot compare these type, even they are constructed typed

Reply:
Yes, it seem that we cannot compare constructed types. Because it is unable to create an expression evaluator for expression


Caused by: org.apache.flink.table.planner.codegen.CodeGenException: Comparable type expected, but was 'ROW<`f0` BOOLEAN, `f1` DATE> NOT NULL'.
	at org.apache.flink.table.planner.codegen.CodeGenUtils$.requireComparable(CodeGenUtils.scala:399)
and because it will call this function

public static boolean isComparable(LogicalType type) {
        return !isRaw(type) && !isMap(type) && !isMultiset(type) && !isRow(type) && !isArray(type);
    }
so we cannot compare these type, even they are constructed typed

Reply:
You don't need to copy the same reply to multiple answers 😅  It's fine if you 
* either skip it or 
* paste a link to the other response or 
* write replied in other question.

Reply:
ok